packages/xo-server-load-balancer/src/plan.js

```diff
@@ -276,6 +278,189 @@ export default class Plan {
+          ) {
+            const source = idToHost[sourceHost.id]
+            const destination = idToHost[destinationHost.id]
+            debugVcpuBalancing(
+              `Migrate VM (${vm.id} "${vm.name_label}") with ${vm.CPUs.number} vCPU to Host (${destinationHost.id} "${destination.name_label}") from Host (${sourceHost.id} "${source.name_label}").`
+            )
+            // 3. Update tags and averages.
+            // This update can change the source host for the next migration.
+            sourceHost.vcpuCount -= vm.CPUs.number
+            destinationHost.vcpuCount += vm.CPUs.number
+
+            const destinationAverages = hostsAverages[destinationHost.id]
+            const vmAverages = vmsAverages[vm.id]
+
+            destinationAverages.cpu += vmAverages.cpu
+            destinationAverages.memoryFree -= vmAverages.memory
+
+            delete sourceHost.vms[vm.id]
+
+            // 4. Migrate.
+            const sourceXapi = this.xo.getXapi(source)
+            promises.push(
+              this._concurrentMigrationLimiter.call(
+                sourceXapi,
+                'migrateVm',
+                vm._xapiId,
+                this.xo.getXapi(destination),
+                destination._xapiId
+              )
+            )
+            debugVcpuBalancing(`VCPU count per host: ${inspect(hostList, { depth: null })}`)
+
+            // 5. Check if source host is still overloaded and if destination host is still underloaded
+            deltaSource = sourceHost.vcpuCount - sourceHost.cpuCount * idealVcpuPerCpuRatio
+            deltaDestination = destinationHost.vcpuCount - destinationHost.cpuCount * idealVcpuPerCpuRatio
+            if (deltaSource < 1) {
+              // eslint-disable-next-line no-labels
+              break destinationLoop
+            }
+            if (deltaDestination >= 0) {
+              break
+            }
+            delta =
+              Math.abs(deltaSource) > Math.abs(deltaDestination)
+                ? Math.ceil(Math.abs(deltaDestination))
+                : Math.ceil(Math.abs(deltaSource))
+          }
+        }
+      }
+    }
+    return Promise.allSettled(promises)
+  }
+
+  _getVCPUHosts(hosts, vms) {
+    const idToHost = {}
+    for (const host of hosts) {
+      const taggedHost = (idToHost[host.id] = {
+        id: host.id,
+        poolId: host.$poolId,
+        cpuCount: parseInt(host.CPUs.cpu_count),
+        vcpuCount: 0,
+        vms: {},
+      })
+
+      // Hide properties when util.inspect is used.
+      Object.defineProperties(taggedHost, {
+        poolId: { enumerable: false },
+        vms: { enumerable: false },
+      })
+    }
+
+    for (const vm of vms) {
+      const hostId = vm.$container
+      if (!(hostId in idToHost)) {
+        continue
+      }
+
+      const host = idToHost[hostId]
+      host.vcpuCount += vm.CPUs.number
+
+      if (vm.xenTools && vm.tags.every(tag => !this._antiAffinityTags.includes(tag))) {
```

Comment:
Yes, I don't try to migrate VMs with anti-affinity tags. It may indeed be a problem if the majority of hosts have an anti-affinity tag, but I don't think that case will be frequent. Maybe if we come up with an idea of a better handling of anti-affinity, we should also modify the way it is handled in the performance plan.